### PR TITLE
Remove $t() from props default

### PR DESCRIPTION
**Pull request by distantnative**
## Describe the PR

In Vue 3, prop default factory functions are losing access to `this`. As a first step to prepare for the change, this PR removes all `this.$t()` calls from the default functions.

## Breaking changes

None

## Related issues/ideas

- Helps with https://github.com/getkirby/kirby/issues/3189

## Ready?

- [x] ~~Unit tests for fixed bug/feature~~
- [x] ~~In-code documentation (wherever needed)~~
- [x] CI checks pass

## When merging

- [ ] Add to [website docs release checklist](https://github.com/getkirby/getkirby.com/pulls) (if needed)
- [ ] Add changes to release notes draft in Notion

### CODE REVIEW

**Review by lukasbestle — COMMENTED** [review]
I think the new behavior could lead to unexpected results in weird edge-cases.

Alternative idea: We could create a new type like `TranslationString` that takes the translation key. All code that interprets the prop value can then know whether it's a normal or translation string. Not sure of the implications, just throwing it into the room. :)

**Comment by distantnative** [comment]
I like that idea. Let's see if Vue does too

**Comment by distantnative** [comment]
Ok maybe not - cause all our props are meant to be strings. Users would usually pass strings not translation keys. Only our defaults use translation keys

**Comment by lukasbestle** [comment]
But couldn't the props look something like this:

```js
placeholder: {
  type: [String, TranslationString],
  default: new TranslationString("email.placeholder")
}
```

**Comment by distantnative** [comment]
And then we could use `TranslationString.toString()` to make the conversion

**Comment by lukasbestle** [comment]
Yep, that sounds good. It could be `this.prop.toString(this)` so the `TranslationString` has access to `this.$t`. If `this.prop` is already a string, I think the passed object for `.toString()` is just ignored by JS.

**Comment by distantnative** [comment]
@lukasbestle Implemented as discussed.

Still feels a bit clumsy in places. Mainly about passing `this` to `toString`. But not sure if we can make this better.

File: panel/src/components/Forms/Input/TextInput.vue [hunk 1]
```diff
@@ -87,6 +87,9 @@ export default {
   computed: {
     direction() {
       return direction(this);
+    },
+    placeholderLabel() {
+      return this.placeholder ? this.placeholder.toString(this) : null;
```
Comment (lukasbestle) [inline panel/src/components/Forms/Input/TextInput.vue]: The `placeholder` prop in this component is defined as `String` only. Have I misunderstood something?

File: panel/src/config/i18n.js [hunk 2]
```diff
@@ -1,9 +1,29 @@
...
+ * TranslationString prop type
+ */
+export class TranslationString {
+  constructor(key) {
+    this.key = key;
+  }
+  toString(app) {
+    return app.$t(this.key);
+  }
+}
```
Comment (lukasbestle) [inline panel/src/config/i18n.js]: For maximum compatibility (e.g. `.length`):

```suggestion
export class TranslationString extends String {
  toString(app) {
    const value = this.valueOf();
    return app ? app.$t(value) : value;
  }
}
```

**Comment by distantnative** [comment]
@lukasbestle I had another idea: use `window.panel.app` to avoid having to call the TranslationString differently than normal strings. Seems to work in my setup - can you confirm?

File: panel/src/config/i18n.js [hunk 3]
```diff
@@ -1,9 +1,28 @@
 
 import str from "../helpers/string.js";
 
+/**
+ * TranslationString prop type
+ */
+ export class TranslationString extends String {
```
Comment (lukasbestle) [inline panel/src/config/i18n.js]:
```suggestion
export class TranslationString extends String {
```

File: panel/src/components/Dialogs/RemoveDialog.vue [hunk 4]
```diff
@@ -15,9 +16,9 @@ export default {
       default: "trash"
     },
     submitButton: {
-      type: [String, Boolean],
-      default() {
-        return this.$t("delete");
+      type: [String, TranslationString, Boolean],
+      default() {
```
Comment (lukasbestle) [inline panel/src/components/Dialogs/RemoveDialog.vue]:
```suggestion
      default() {
```

**Review by lukasbestle — COMMENTED** [review]
That's even better, yes!

I think we can even get rid of the `[String, TranslationString]` type definitions for the props now. If Vue checks the types using `prop instanceof String`, it should still work with a simple `String` because our new class extends that. I haven't tested how Vue behaves, but it would make it all even simpler.

Besides that, two minor whitespace issues. Otherwise the PR is great! 😍

**Comment by distantnative** [comment]
@lukasbestle that does seem to work!

Love PRs like this where we start at "works but meh" and add onto each other's ideas to end at "that's quite a neat solution".

**Review by lukasbestle — APPROVED** [review]

**Comment by lukasbestle** [comment]
> Love PRs like this where we start at "works but meh" and add onto each other's ideas to end at "that's quite a neat solution".

💯

**Comment by bastianallgeier** [comment]
Maybe I'm missing something obvious, but we still need to import the TranslationString class in every instance that needs translated props. IMHO this is the biggest issue with this rewrite and the changes in Vue 3. Components in plugins cannot simply import it and can no longer have translated props out of the box. 

I'm also not 100% sure I understand why you use the TranslationString thingy instead of importing Vue and using Vue.$t()

**Comment by distantnative** [comment]
I think you're getting it a bit wrong: you only need to import TranslationString when you cannot use `this.$t` - which really only concern factory function or prop default values. Passing prop values will be still fine with `this.$t` wherever `this` is available in context. Just default factory functions won't have that anymore.

The alternative is to create a lot of computed properties that check if the prop value is empty and either pass the prop value or the default instead. We went for the TranslationString way since it doesn't create so much more markup.

Importing Vue isn't a thing anymore in Vue 3. There is no global Vue. So since this change is mainly to fix/prepare one aspect for Vue 3, of course I didn't went for another way not supported by Vue 3.

**Comment by bastianallgeier** [comment]
ok, but this still means that a plugin dev cannot provide translated default values for props anymore. When we start trying to think more like plugin devs and their major problems in components, we need a fix that might work for them too. They won't be able to import TranslationString at all.

**Comment by distantnative** [comment]
I don't think blowing up our code instead would be a solution for them either. It's a really annoying change of Vue 3 - one of many - that we all will have to adapt to. We could try to expose TranslationString somehow to plugins developers but that will be rather difficult I am afraid since plugins run first before all of our code.

**Comment by lukasbestle** [comment]
Plugin devs have multiple alternative options:

- They can use props that are just the translation key. Often times, it's not necessary to have the flexibility of a fixed or translation string in plugins as they have much more control about their configuration.
- Or they can use a computed prop that automatically fills in a translation string default value if the prop is not set.
- Or they can design their props so that they always take a fixed string that can then be dynamically populated with a translation string from the parent component.

I agree with Nico that there's likely no good general solution that we can implement for plugin devs. We don't have the Vue 3 changes under control, so it requires that everyone adapts.

**Comment by bastianallgeier** [comment]
Ok, I'm convinced :) We also have a nice grace period if plugin devs still need it.

**Pull request merged**